### PR TITLE
intel/adsp: PM fixes

### PR DESCRIPTION
**Pull request by ceolin**
Several power management fixes, e.g:

- Remove unnecessary cache flushes
- Avoid complicated logic just removing invalid power states from secondary cores
- Remove duplicated cache logic

See individual commit messages for more details.

### CODE REVIEW

File: soc/intel/intel_adsp/ace/power.c [hunk 1]
```diff
@@ -135,7 +135,13 @@ static ALWAYS_INLINE void _save_core_context(uint32_t core_id)
 	core_desc[core_id].thread_ptr = XTENSA_RUR("THREADPTR");
 	__asm__ volatile("mov %0, a0" : "=r"(core_desc[core_id].a0));
 	__asm__ volatile("mov %0, a1" : "=r"(core_desc[core_id].a1));
+
+#if CONFIG_MP_MAX_NUM_CPUS == 1
+	/* With one core only, the memory is mapped in cache and we need to flush
+	 * it.
```
Comment (andyross) [inline soc/intel/intel_adsp/ace/power.c:135]: Actually I'm pretty sure this is needed.  At least on early cAVS platforms, the linker script would only do the memory space separation on multicore setups, leaving .bss data like core_desc[] cached.  Though not sure if that was MP_MAX_NUM_CPUS or SMP or what.

Basically the comment sounds to me like someone bumped into this as a bug and fixed it?

**Review by andyross — COMMENTED** [review]
One note to check in the linker script.

**Comment by marc-hb** [comment]
Please wait for https://github.com/thesofproject/sof/pull/8968 to be merged first and then submit this to SOF CI with a throw-away, draft `sof/west.yml` PR. If nothing else this is really the area that MUST be tested by SOF.

Thanks for the cleanups but we're just exiting a very long and painful PM tunnel and we _really_ don't want to enter a new one.

- https://github.com/zephyrproject-rtos/zephyr/pull/70184#issuecomment-2000467898

If it's not tested then it does not work (c)

File: soc/intel/intel_adsp/ace/power.c (hunk 1 above)
Comment (ceolin) [reply soc/intel/intel_adsp/ace/power.c:135]: it was always flushing it, but this memory is not in cached region ... unless we have only one core.
Basically, I had removed this flush but when we built with only one core the memory is mapped in cache and we need it.

The change is basically doing, call this cache flush *if* we build with only one core.

**Comment by ceolin** [comment]
> Please wait for [thesofproject/sof#8968](https://github.com/thesofproject/sof/pull/8968) to be merged first and then submit this to SOF CI with a throw-away, draft `sof/west.yml` PR. If nothing else this is really the area that MUST be tested by SOF.
> 
> Thanks for the cleanups but we're just exiting a very long and painful PM tunnel and we _really_ don't want to enter a new one.
> 
> * [soc: intel_adsp_cavs: store PS when power gating secondary core #70184 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/70184#issuecomment-2000467898)
> 
> If it's not tested then it does not work (c)

I know about it, and that is the reason I have limited this pr to just small changes. I have a couple of more complex changes waiting to be submitted.

I have no problem waiting, lets put a DNM label here.

**Comment by marc-hb** [comment]
> but we're just exiting a very long and painful PM tunnel ...

Apparently not yet:
- #70494

**Review by kv2019i — DISMISSED** [review]
Test run with SOF pull-requqest CI would be good before merge, but code looks like a welcome improvement and seems ready to go.

**Review by tmleman — COMMENTED** [review]
I've been using these changes locally and haven't encountered any issues (without asserts enabled, so as @kv2019i suggested, I'd like to see them run through SOF CI).

**Comment by teburd** [comment]
Feel free to steal and pilfer the sample app I wrote to ensure things work and stay working, would be really nice to see that included here in some form of test.

**Review by kv2019i — APPROVED** [review]
Looks good now. SOF side tests show no regression (see test results at https://github.com/thesofproject/sof/pull/9126 ).

**Review by nashif — APPROVED** [review]

**Review by tmleman — APPROVED** [review]

**Comment by marc-hb** [comment]
Is this expected? I didn't look at the code change sorry

https://github.com/thesofproject/sof/actions/runs/9100669711/job/25016041801?pr=9126

```
/zep_workspace/sof/src/audio/base_fw_intel.c:132:39: error: implicit declaration of function 'HPSRAM_REGS' [-Werror=implicit-function-declaration]
  132 |                 tuple_data[index++] = HPSRAM_REGS(i)->HSxPGCTL;
      |                                       ^~~~~~~~~~~

```

**Comment by ceolin** [comment]
> Is this expected? I didn't look at the code change sorry
> 
> [thesofproject/sof/actions/runs/9100669711/job/25016041801?pr=9126](https://github.com/thesofproject/sof/actions/runs/9100669711/job/25016041801?pr=9126)
> 
> ```
> /zep_workspace/sof/src/audio/base_fw_intel.c:132:39: error: implicit declaration of function 'HPSRAM_REGS' [-Werror=implicit-function-declaration]
>   132 |                 tuple_data[index++] = HPSRAM_REGS(i)->HSxPGCTL;
>       |                                       ^~~~~~~~~~~
> ```

The code is not touching this. This definition is inside `adsp_memory.h` header.

File: soc/intel/intel_adsp/ace/power.c [hunk 2]
```diff
@@ -334,25 +334,26 @@ void pm_state_set(enum pm_state state, uint8_t substate_id)
 			/* do power down - this function won't return */
 			ret = pm_device_runtime_put(INTEL_ADSP_HST_DOMAIN_DEV);
 			__ASSERT_NO_MSG(ret == 0);
-			power_down(true, uncache_to_cache(&hpsram_mask),
+			power_down(true, sys_cache_cached_ptr_get(&hpsram_mask),
```
Comment (marc-hb) [inline soc/intel/intel_adsp/ace/power.c:337]: @ceolin , @lyakh this fails a sparse check:

https://github.com/thesofproject/sof/actions/runs/9253499236/job/25453295192?pr=9168

```
zephyr/soc/intel/intel_adsp/ace/power.c:337:66: warning:
    incorrect type in argument 2 (different address spaces)
```

**Pull request merged**